### ISSUE TITLE
Require valid custom DNS address before enabling the 'Use custom DNS' setting

### ISSUE THREAD

**Issue by data-sync-user** [mozilla-mobile/mozilla-vpn-client#5849]
This ticket is carried over from [https://mozilla-hub.atlassian.net/browse/VPN-3891.](https://mozilla-hub.atlassian.net/browse/VPN-3891.|smart-link) 

**Acceptance Criteria:**

* Opting into “Custom DNS” only goes into effect when the user has selected the appropriate radio button AND has inputted a valid DNS address in the text field. Not meeting these conditions means that the change has not happened and none of the logic specified above gets triggered (no silent server switching, no server disconnection and reconnection, no verification against privacy features, etc). 
* The validation happens in the following situations: 
  * When the user taps/clicks ‘out’ of the text field. 
  * When the user taps/clicks ‘back’ on the screen. 
  * When the user taps/clicks the ‘home’, ‘messages’ or ‘settings’ navbar menu items. 
  * The validation doesn’t happen when the user selects the ‘Custom DNS', inputs a valid DNS server location, and (without any of the above situations), simply quits the app. In this case, navigating back to the ‘DNS Settings’ screen shows whatever were the previous settings
* The SUMO article about Custom DNS (Created via [[https://bugzilla.mozilla.org/show_bug.cgi?id=1810367%7Chttps://bugzilla.mozilla.org/show_bug.cgi?id=1810367%7Csmart-link](https://bugzilla.mozilla.org/show_bug.cgi?id=1810367%7Chttps://bugzilla.mozilla.org/show_bug.cgi?id=1810367%7Csmart-link|smart-link)  ]) should be updated to reflect that the VPN needn’t be off to change it

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/VPN-4070)
┆Reporter: Lesley Norton

**Comment by data-sync-user** [comment]
➤ Lesley Norton commented:

It doesn’t need to block this ticket but I think it would be helpful to have some additional messaging here to inform the user that their change will not take effect until a valid custom DNS has been entered.

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Verified this on the latest 2.14 build, across platforms, and the following were observed:

  * the silent server switching / server disconnection and reconnection logics / verification against privacy features are triggered when selecting the “Use custom DNS” radio button, and not only after entering a valid DNS address in the text field. 
  * if leaving the custom dns field empty / entering an invalid dns address, when navigating away and returning, the “Use standard DNS” radio button is selected. 
  * while entering a valid custom dns, the address is validated while the field is still active (no need to click outside / click “back” / navigate away / quit from tray bar to apply changes).
  * “Back” option triggers different behavior than clicking outside / using the nav bar → we logged 2 separate issues for this: https://mozilla-hub.atlassian.net/browse/VPN-4329 ( https://mozilla-hub.atlassian.net/browse/VPN-4329|smart-link )  / https://mozilla-hub.atlassian.net/browse/VPN-4330 ( https://mozilla-hub.atlassian.net/browse/VPN-4330|smart-link ) 

Based on our findings we are reopening this ticket.

**Comment by data-sync-user** [comment]
➤ Andrea Marchesini commented:

1. the extra silent-switch → I cannot reproduce it. Can you file a separate issue please?
  2. Santiago Andrigo this sounds like the right behavior. Am I right?
  3. Santiago Andrigo also this sounds correct. The validation happens at runtime, and not when the settings are applied.
  4. Working on the other 2 bugs.

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Andrea Marchesini, of course. Tracking issue: https://mozilla-hub.atlassian.net/browse/VPN-4331 ( https://mozilla-hub.atlassian.net/browse/VPN-4331|smart-link )

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

If point 3 is correct, then the AC needs to be updated. 

Also, regarding the “no verification against privacy features” part → the overriding of the privacy features modal is displayed when selecting the “Use custom DNS” option, without entering an address in the custom field, which seems the correct behavior, but from the AC, it seems that it should be displayed only after entering a valid DNS address, Santiago Andrigo / Andrea Marchesini ?

**Comment by data-sync-user** [comment]
➤ Santiago Andrigo commented:

Regarding point 3 ---> my intention was that validation happened in the situations mentioned above in this ticket. It seems like we are doing validation and application all the time, which creates problems: https://mozilla-hub.atlassian.net/browse/VPN-3428 ( https://mozilla-hub.atlassian.net/browse/VPN-3428|smart-link ). Andrea Marchesini maybe better to implement as per above?

**Comment by data-sync-user** [comment]
➤ Andrea Marchesini commented:

Santiago Andrigo the validation happens at editing time. The setting change happens when the field looses focus or visibility.

That bug is a qt issue for Mac only. 

If we don't do as implemented, the user has to move away from the text field to see if the input is valid.

**Comment by data-sync-user** [comment]
➤ Valentina Virlics commented:

Verified all mentioned acceptance criteria, while using the last 2.14.0 (2.202303201325) build, across platforms.

  1. The “When the user taps/clicks ‘out’ of the text field.“ AC is not fulfilled. Andrea Marchesini Santiago Andrigo Is this expected?
  2. Please also see Bianca’s question, below - https://mozilla-hub.atlassian.net/browse/VPN-4070?focusedCommentId=658837 ( https://mozilla-hub.atlassian.net/browse/VPN-4070?focusedCommentId=658837|smart-link ) 

Thank you!

**Comment by data-sync-user** [comment]
➤ Santiago Andrigo commented:

Regarding Bianca’s question: 

Yes, the intended behavior is that the modal shows up as soon as the user clicks “Custom DNS”, even if no IP is entered. This ticket was created as an offshoot of https://mozilla-hub.atlassian.net/browse/VPN-3891 ( https://mozilla-hub.atlassian.net/browse/VPN-3891|smart-link ), where this was clarified.

Regarding Valentina’s question:

Baku and I agreed on this AC offline, so would you mind filing a bug for it?Andrea Marchesini A refinement we can introduce is that the DNS does goes into effect upon ‘Quit’ (but not upon Kill), contrary to what my original proposal was. What do you think?

**Comment by data-sync-user** [comment]
➤ Bianca Hidecuti commented:

Verified this on the latest 2.14 builds, across platforms and now the AC is met.

We will leave this open until we will be able to verify it on the new RCs as well.

Adding the 2.14 fix version as well.
